troubleshoot-mesheryoperator (#9479)
meshery/meshery: docs/pages/reference/troubleshooting/meshery-operator-meshsync-troubleshooting.md

permalink: guides/troubleshooting/meshery-operator-meshsync
language: en
abstract: This documentation provides comprehensive guidance on troubleshooting in Meshery Operator, MeshSync and Broker, ensuring you can address common issues efficiently.
type: guides
category: troubleshooting
---

{% include alert.html type="dark" title="Meshery Error Code Reference" content="Have specific error with an error code? See the <a href='/reference/error-codes'>Meshery Error Code Reference</a> for probable cause and suggested remediations." %}

There are common issues Meshery users may face whiile operating the [Meshery Operator]({{site.baseurl}}/concepts/architecture/operator/) and its custom controllers, [MeshSync]({{site.baseurl}}/concepts/architecture/meshsync) and [Broker]({{site.baseurl}}/concepts/architecture/broker), that can be resolved by performing specific actions. This documentation aims to empower users by providing a set of troubleshooting tools and actions.

## Understanding the Status of Meshery Operator, MeshSync, and Meshery Broker

The following table describes the various states of MeshSync and Meshery Broker and their implications.

**MeshSync:**

> - **NOT ACTIVE:** Custom Resource not present.

Reviewer: ```suggestion
```
Custom Resource are not present, it is Undeployed.

- **ENABLED:** Custom Resource present. MeshSync Controller is not connected to Broker.

> - **DEPLOYED:** Custom Resource present. MeshSync Controller present and healthy (what does healthy mean)?

Reviewer: healthy meaning, the status.phase of the pod is "Running".

Author: @MUzairS15 please address this in the docs.

> - **DEPLOYED:** Custom Resource present. MeshSync Controller present and healthy (what does healthy mean)?

Reviewer: ```suggestion
- **DEPLOYED:** Custom Resource present. MeshSync Controller is present but the state is not RUNNING or ERRDISABLE, though
```

- **RUNNING:** MeshSync pod present and in a running state.
- **CONNECTED:** Deployed and connected to Broker.

> - **NOT ACTIVE:** Custom Resource not deployed or no NATS connection.

Reviewer: ```suggestion
```

- **UNDEPLOYED:** Custom Resource not present.

**Meshery Broker:**

> - **DEPLOYED:** Custom Resource deployed, External IP exposed.

Reviewer: ```suggestion
- **DEPLOYED:** External IP not exposed OR External IP exposed but Meshery Server is not connected as a client to Broker hence data is not being published.

```

> - **UNDEPLOYED:** Custom Resource deployed or External IP not exposed.

Reviewer: ```suggestion
- **UNDEPLOYED:** Custom Resource not deployed.
```

### In-Cluster Deployment

Meshery Operator, MeshSync, and Broker are deployed in the same cluster as Meshery Server. This is the default deployment scenario when using `mesheryctl system start` or `make run-local`.

### Out-of-Cluster Deployment

1. Meshery Server is deployed on any Docker host (- Meshery Server is deployed on a Docker host, and Meshery Operator is deployed on a Kubernetes cluster).
_or_
2. Meshery is managing multiple clusters, some of which are not the cluster unto which Meshery Server is deployed.

## Fault Scenarios

Common failure situations that Meshery users might face are described below.

1. No deployment of Meshery Operator, MeshSync, and Broker.
    1. Probable cause: Meshery Server cannot connect to Kubernetes cluster; cluster unreachable or kubeconfig without proper permissions needed to deploy Meshery Operator; Kubernetes config initialization issues.
1. Meshery Operator with MeshSync and Broker deployed, but Meshery Server is not receiving data from MeshSync or data the [Meshery Database]({{site.baseurl}}/concepts/architecture/database) is stale.

>     1. Probable cause: Meshery Server lost subscription to Meshery Broker; Broker server not expoerting external IP; MeshSync not connected to Broker; MeshSync not running; Meshery Database is stale.

Reviewer: ```suggestion
    1. Probable cause: Meshery Server lost subscription to Meshery Broker; Broker server not exposed to external IP; MeshSync not connected to Broker; MeshSync not running; Meshery Database is stale.
```
